### PR TITLE
Add Session.purpose field and supporting models

### PR DESCRIPTION
**Pull request by jennifer-richards**
* Add SessionPurposeName for enumerating options
    - each has a list of allowed TimeSlotTypeNames
  * Add Session.purpose field
  * Add GroupFeatures.session_purposes field to enable
    purposes for different session types
  * Add a JSONForeignKeyListValidator to validate that
    a JSONField is a valid list of foreign keys

The contents of the populate_* migrations are approximate
and will likely be changed as we move forward.

### CODE REVIEW

File: ietf/utils/validators.py [hunk 1]
```diff
@@ -248,3 +249,38 @@ def clean(self, x):
             return x.width, x.height
         except FileNotFoundError:
             return 0, 0  # don't fail if the image is missing
+
+
+@deconstructible
```
Comment (meadmaker) [inline ietf/utils/validators.py:254]: Doesn't this require a `@postmodernism` decorator as well?

**Review by meadmaker — COMMENTED** [review]
I don't see any incorrect lines of code in this, so there's an urge to "Approve" this.

However, storing relational data in a JSON string within the database is a big red flag for me.  All of that could be accomplished fairly easily using additional tables and regular foreign keys.  Why have you chosen to approach the problem with JSON instead?

**Comment by jennifer-richards** [comment]
> However, storing relational data in a JSON string within the database is a big red flag for me. All of that could be accomplished fairly easily using additional tables and regular foreign keys. Why have you chosen to approach the problem with JSON instead?

Yes, I agree it's a red flag. [I discussed it with Robert on the IETF notes site](https://notes.ietf.org/#Constraining-types) (First and third paragraph are Robert, middle is me). The original suggestion was a many-to-many field, which got some pushback.

As I look at it now, I think some of the appearance of complexity is because I called out the "through" models. That's really not needed for a simple many-to-many field. I will give it a stab with that and discuss with Robert.

If you don't object, I'd like to merge this to avoid having too many live branches of tix3330, then refactor in a future chunk.

**Comment by jennifer-richards** [comment]
Per our discussion, merging this as-is to keep the spice flowing.

**Pull request merged**